Manual Web User Deactivation

minhaminha opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

This PR adds the ability to manually deactivate and reactivate web users, exactly the same way as you would mobile workers. Not to be confused with the existing way to deactivate users that use SSO - this deactivation (going to refer to as domain active status) is scoped only to a single domain. Deactivated web users will be:

- Unable to log in or access the deactivated domain
- Not counted towards the total web user count for billing purposes
- Omitted from the recipient list of any scheduled reports/broadcasts/conditional alerts they might've been part of

Additionally, you'll now be able to filter any reports with user type filtering by active vs deactivated web users (this includes web user download). For more things that deactivation affects, please see the design doc below or the commit names. 

![screen 1](https://github.com/user-attachments/assets/bc02a3e7-a982-401a-a523-616ed8f85fec)

![screen 2](https://github.com/user-attachments/assets/8c7bc2af-6058-4d05-9ed3-131e51274ad8)

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->

[Jira epic](https://dimagi.atlassian.net/browse/USH-5213)
[Design Doc](https://docs.google.com/document/d/1chcxTrcjvu-hekJPw7f_pjUnfR10vyTRHpcjmYizu2Q/edit?tab=t.0#heading=h.ipgou4bu0qoj) / [Tech Spec](https://docs.google.com/document/d/1lmvuvadq0Lw7iAg2areB-w5kSz1ShZ9nLy8pjtTRgrI/edit?tab=t.0)

The deactivation state will live in the individual domain_memberships obj a web user has. See [this PR](https://github.com/dimagi/commcare-hq/pull/36397) to see how that'll be queried in ES. 

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

This will not be behind a FF (as specified in the design doc) 

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

Tested locally and on staging. 

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

--

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

[QA Ticket](https://dimagi.atlassian.net/browse/QA-7746)

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

minhaminha commented:
FYI: All the test failures are because `DomainMemebership` currently lacks the `is_active` attribute, which is added on the [other PR](https://github.com/dimagi/commcare-hq/pull/36397)

mkangia commented:
Can web users not submit data via phone application? I thought they could. What stops them?

mkangia commented:
@minhaminha 
That is my understanding. You can just login a mobile app using a web user account, it will only fail in one known scenario.
Should be easy to test.

Jtang-1 reviewed (commented):
Got through reviewing 9e84ce24b02dc6354558280aad16a528342befef—I'll continue with the rest tomorrow. This looks like a pretty involved change, and you did a great job uncovering all the little places that needed to be touched. Nice work!

minhaminha commented on corehq/apps/users/models.py:
```diff
@@ -748,6 +748,12 @@ class MultiMembershipMixin(_AuthorizableMixin):
 
+    @memoized
+    def is_active_in_domain(self, domain):
+        if self.get_domain_membership(domain):
```
yep it's all in the user doc but true this is better

minhaminha commented on corehq/apps/users/dbaccessors.py:
```diff
@@ -240,6 +253,18 @@ def get_web_user_count(domain, include_inactive=True):
+        start_date = today - relativedelta(months=1)
+        for u in web_users:
+            if not u.is_active_in_domain(domain):
+                user_history = UserHistory.objects.filter(
```
hmm that's why I put it under the `exclude_deactivated_web` conditional, which is only set to True from accounting contexts. Do you think that's not enough? Maybe a comment to explain it's purpose?

minhaminha commented on corehq/apps/users/dbaccessors.py:
```diff
@@ -240,6 +253,18 @@ def get_web_user_count(domain, include_inactive=True):
+        today = datetime.datetime.today()
+        start_date = today - relativedelta(months=1)
+        for u in web_users:
+            if not u.is_active_in_domain(domain):
```
1. I think I wasn't sure how to do that which is why I'm filtering like this but I can look into that again 
2. inactive in this case refers to the user level is_active status. For web users, this is used to refer to the fact that they use SSO but had their access revoked by their idp. For mobile workers, it refers to their domain level deactivated/active state. Exclude deactivated web refers only to domain level deactivated/active state for web users.

Jtang-1 replied:
Thanks for the clarifying definitions. Before you look into the view/index again, I would want to get @esoergel thoughts if it's even a concern. Do you think think doing the filtering in memory here is fine?

MartinRiese commented on corehq/apps/domain/decorators.py:
```diff
@@ -107,6 +107,9 @@ def call_view():
         if couch_user.is_member_of(domain_obj, allow_enterprise=True):
+            if (couch_user.is_web_user() and not couch_user.is_superuser
+               and not couch_user.is_active_in_domain(domain_name)):
+                return HttpResponseForbidden(_("You have been deactivated from {domain}.").format(domain=domain))
```
I think this shows up as just a text in the browser. I wonder, if we can make that a templateResponse like two lines down. So it looks better.

minhaminha replied:
Added a new `deactivated_notice.html` template to show instead of a blank page with text

MartinRiese commented on corehq/apps/domain/views/base.py:
```diff
@@ -68,9 +68,16 @@ def select(request, do_not_redirect=False, next_view=None):
                 or domain_obj.is_snapshot
             ):
+                redirect_last_domain = True
+                if (
```
That took me a moment to understand. But I agree. I am reading the code the same way.

MartinRiese commented on corehq/apps/enterprise/api/resources.py:
```diff
@@ -274,7 +274,8 @@ class WebUserResource(ODataEnterpriseReportResource):
         'last_access_date': 4,
         'status': 5,
-        'domain': 6,
+        'is_active_in_domain': 6,
```
Curious why you are reordering the columns?

minhaminha replied:
This was specifically for tests (only tests seem to be using this column map?) which expected a certain order that was impossible to do with is_active_in_domain being 7

MartinRiese commented on corehq/apps/hqwebapp/views.py:
```diff
@@ -560,8 +560,12 @@ def logout(req, default_domain_redirect='domain_login'):
     if referer and domain:
-        domain_login_url = reverse(default_domain_redirect, kwargs={'domain': domain})
-        return HttpResponseRedirect('%s' % domain_login_url)
+        show_domain_login = True
+        if req.couch_user.is_web_user() and not req.couch_user.is_active_in_domain(domain):
+            show_domain_login = False
```
`show_domain_login = not (req.couch_user.is_web_user() and not req.couch_user.is_active_in_domain(domain))`?

minhaminha replied:
updated - clearly I was not in the headspace for understanding conditionals this day

MartinRiese reviewed (commented):
Halfway. Just submitting so you get the comments. Will continue shortly

MartinRiese commented on corehq/apps/user_importer/importer.py:
```diff
@@ -816,6 +819,13 @@ def _modify_existing_user_in_domain(self, membership, role_qualified_id,
+        if str(current_user.is_active_in_domain(self.domain)) != imported_is_active:
+            if imported_is_active == 'True':
+                current_user.reactivate(self.domain, changed_by=web_user_importer.upload_user)
+            elif imported_is_active == 'False':
```
Can this not just be an `else`?

minhaminha replied:
I think I wanted users to be very intentional about deactivating users through the bulk upload. I don't want deactivation to happen if they accidentally leave it blank, misspell 'True', or whatever else.

Jtang-1 commented on corehq/apps/users/dbaccessors.py:
```diff
@@ -183,6 +191,11 @@ def _get_invitations_by_filters(domain, user_filters, count_only=False):
     Ignores any other filters.
     """
+    only_active = user_filters.get("user_active_status", None)
+    if not only_active and only_active is not None:
```
Is this conditional right? I'm interpreting `user_active_status=True` means returning only invitations whose users are active on the domain (which would be 0). But that conditional would evaluate as False

MartinRiese replied:
not sure if I understand this. can a user for an invitation ever be active on that same domain?

Jtang-1 commented on corehq/apps/users/views/__init__.py:
```diff
@@ -957,6 +986,28 @@ def undo_remove_web_user(request, domain, record_id):
+@location_safe
+def deactivate_web_user(request, domain, couch_user_id):
+    user = WebUser.get_by_user_id(couch_user_id, domain)
+    user.deactivate(domain, changed_by=request.couch_user)
```
I may have missed it - is there any validation that prevents the user from deactivating themselves? If not, I think that can be added here and then on the frontend, don't display the "deactivate" button for the request user.

minhaminha replied:
The deactivate/reactivate buttons already don't show/aren't initialized for the current user.

esoergel committed:
Remove old once-off migration

esoergel committed:
Make is_admin fully available filter

esoergel committed:
Combine domain and domains filters

esoergel committed:
Make domain filter handle is_active status

MartinRiese committed:
Fix usage of UserES

Move include inactive into domain call

MartinRiese committed:
Fix usage of UserES

Add include_inactive to active_user_es_query so it can pass it to domain

MartinRiese committed:
Fix UserES usage

Straight forward replacement

MartinRiese committed:
Fix UserES usage

is_active is covered by domain now. By default it only includes
active users

esoergel committed:
Add is_active field to domain memberships

esoergel committed:
Remove default activity filters

MartinRiese committed:
Fix UserES usage

MartinRiese committed:
Merge branch 'es/is_active_query' of github.com:dimagi/commcare-hq into es/is_active_query

MartinRiese committed:
Fix UserES usage: User active on any domain

For mass emails we just care that the user is
active on at least one domain

MartinRiese committed:
Fix lint errors

MartinRiese committed:
Merge branch 'master' into es/is_active_query

MartinRiese committed:
Fix tests

MartinRiese committed:
Fix UserES usage

Domain now filters out inactive users by default
Add right flags to get intended users.

MartinRiese committed:
Fix UserES usage

Domain now filters out inactive users by default
Add right flags to get intended users.

esoergel committed:
This also includes the demo user by default

esoergel committed:
A few easy fixes

esoergel committed:
Make domain available to locations dbaccessors

MartinRiese committed:
Fix lint error

esoergel committed:
That's the last of 'em

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into ml/web-user-deactivation-manual

esoergel committed:
Merge remote-tracking branch 'origin/es/is_active_query' into ml/web-user-deactivation-manual

MartinRiese committed:
Fix lint issue

MartinRiese committed:
Update renamed label in test

esoergel committed:
Update web users list

esoergel committed:
Remove duplicate domains fn in favor of domain

MartinRiese committed:
Merge branch 'ml/web-user-deactivation-manual' of github.com:dimagi/commcare-hq into ml/web-user-deactivation-manual

MartinRiese committed:
Fix test

functions calls were changed in a previous commit

MartinRiese committed:
Fix test: domain should be passed in the builder

MartinRiese committed:
Remove additional check

The same check already happens in the
requires_privilege_for_commcare_user decorator with
addtion of accounting for superusers. Whereas the removed check
will throw and exception for superusers

MartinRiese committed:
Fix expand mobile worker filter

Fix its usage of the UserES filter to match the new
functionality. Update the test data

MartinRiese committed:
Remove debug statement

MartinRiese committed:
Bootstrap 5 Migration - Rebuilt diffs

esoergel commented on corehq/apps/reports/filters/users.py:
```diff
@@ -351,6 +351,7 @@ def filter_context(self):
     @classmethod
     def user_es_query(cls, domain, mobile_user_and_group_slugs, request_user):
         # The queryset returned by this method is location-safe
+        # import pdb; pdb.set_trace()
```
:eyes:

MartinRiese committed:
Merge branch 'master' into ml/web-user-deactivation-manual

minhaminha committed:
Merge branch 'master' into ml/web-user-deactivation-manual
